README.md
# swrcfit (SWRC Fit) - Fitting soil water retention curve

Choose language.

- [English](doc/en/README.md)
- [日本語](doc/ja/README.md)

This file is written in markdown language and online version is better
for browzing. User's manual of SWRC fit for each version is available
here: https://github.com/sekika/swrcfit/wiki/User%27s-manual


ChangeLog
SWRC Fit Version History

* Notable changes.

Version 3.1 (2018-09-27)

  * Note for Mac users: homebrew/science tap on homebrew is now deprecated.
    Change to sekika/sci tap by
    brew untap homebrew/science
    brew tap sekika/sci
  * BUG FIX: Previous versions of SWRC Fit does not work on Octave 4.4 because
    "duplicate subfunction" error which did not appear in previous versions of
    Octave arises. This bug was fixed.
  * BUG FIX: When normcdf is not available, nan package from octave-forge is
    loaded. Install script was changed to load the nan package.

Version 3.0 (2016-09-10)

  * Added support for Fredlund and Xing model (FX model)
  * Individual model selection with bc, vg, ln, fx, db, bl options
  * Added output options: data, K, r2, rmse, ns, aic, bic
  - Correction function of FX model: fxc, psir, psimax
  - Unimodal and bimodal plots on same figure
  - Added figure option: onemodel (minimum AIC selection)
  - Documents in the wiki were moved to package. This will allow browzing
    documents in each release.
  - Confirmed running on Bash on Windows 10. See document for detail.

Version 2.1 (2015-09-23)

  * BUG FIX: SWRC Fit version 2.0 or before does not work with the latest
    version of Octave. Command "pkg load all" was added to fix this problem,
    to load necessary packages explicitly. In older version of Octave (not
    sure exactly until which version), packages could be loaded automatically
    without this command, but it should be specified explicitly with the
    latest verion of Octave. Therefore, users should upgrade to SWRC Fit
    version 2.1, to use it with the latest version of Octave.
  - Paramter restrictions of alpha < 1/h1 (VG model) and hm > h1 (LN model),
    where h1 is the smallest suction head of the measured point, were imposed,
    because in certain cases inflection points of VG and LN models become too
    small while the saturated water content becomes unrealistically high.
    Some other fixes for parameter handling.

Version 2.0 (2015-06-07)

  * New feature to draw graphs of the result
  * Simple output mode
  * Show version with -v option
  * ./configure; (sudo) make install installs both program and man page.
  * Multiple setting with equation of filename without limitation of numbers
  * Read default setting at $HOME/.swrcfitrc and ./setting.txt
  - Manual in man page style was created.
  - Japanese document (README.ja.md)
  - Install.sh creates destination directory
  - In a certain case reported by a user, "error: weighted residuals are
    not real" was returned from initial calculation of BC fit with leasqr.m.
    Error handling block was added to fix this problem. If you find further
    problem with your data, please report bug to the author.

Version 1.3 (2014-09-07)

  * swrc.m and bimodal.m were combined to swrcfit, and unimodal and bimodal
    calculation is to be selected from setting.
  * Data file and setting file is specified with commandline option.
  - Sample data and result file were added.
  - User's manual was changed to markup file (README.md)
  - Some file names were changed.
  - Install script for Unix-like systems was made.

Version 1.2 (2007-09-02)

  - User's manual (manual.pdf) was made.
  - Correlation matrix and standard deviation is calculated in advanced mode
    of the calculation option.
  - Restiction of residual water content not becoming negative set as default.

Version 1.1 (HESSD version) (2007-02-27)

  * Identical to the program published as a supplementary file of HESSD journal
    (Hydrology and Earth System Sciences Discussions, 4: 407-437, 2007.)
  * BL model was added as a bimodal model (bimodal.m).
  - Option for imposing the restriction of positive value of residual water
    content (pqr) was added.
  - Publication date of the paper is shown. Version 1.0 and 1.1 were uploaded
    to sourforge on 2007-06-05.

Version 1.0 (JJSSP version) (2006-07-28)

  * Identical to the program published in JJSSP journal
    (Journal of Japanese Society of Soil Physics, 105:67-78.)
    Submission date of the paper is shown. Publication date is 2007-03-20.


swrc.m
#!/usr/bin/octave -qf
# SWRC Fit - swrc.m
# URL: http://swrcfit.sourceforge.net/
# Author: Katsutoshi Seki
# Licence: GNU Public License
# Version: 1.2

# Read data from file
load swrc.txt
[s, i] = sort (swrc (:, 1));
sorted = swrc (i, :);
x=sorted(:,1); y=sorted(:,2);
if columns(sorted)==2,
  wt=ones(length(y),1);
else
  wt=sorted(:,3);
end

# Setting
output_precision=7; # precision of the output
qsin = max(y); # initial value of qs
cqs=1; # cqs=1; qs is variable, cqs=0; qs is constant
qrin = min(y); # initial value of qr
cqr=1; # cqr=1; qr is variable, cqr=0; qr is constant
# qrin=0; cqr=0; # For setting qr=0 as a constant
pqr=1; # pqr=1; qr >= 0, pqr=0; qr can be negative
adv=0; # adv=1; advanced output; adv=0; normal output;

# Set initial values
qs=qsin; qr=qrin;
hbi = sum (y > max(y) * 0.95 + min(y) * 0.05)+1;
hb=x(hbi); hby=(y(hbi)-qr)/(qs-qr);
if hb==0, hb=x(2)/10;, end
hli = sum (y > max(y) * 0.15 + min(y) * 0.85);
if hbi==hli; hli=hli+1;, end
hl=x(hli); hly=(y(hli)-qr)/(qs-qr);
lambda=-log(hly/hby)/log(hl/hb);
if lambda < 0.1, lambda=0.1;, end
if lambda > 10, lambda=10;, end

# Brooks and Corey Model
disp ("=== BC model ===");
x2 = x .+ (x == 0) .* 0.001;
function ret=BC(x,p)
  ret = (x > p(3)) .* (p(2)+(p(1)-p(2)).*(x./p(3)).^(-p(4))) + (x <= p(3)) .* p(1);
endfunction
function ret=BC2(x,p)
  ret = (x > p(3)) .* (abs(p(2))+(p(1)-abs(p(2))).*(x./p(3)).^(-p(4))) + (x <= p(3)) .* p(1);
endfunction

pin=[qs, qr, hb, lambda];
stol=1;
dp=[0.00001*cqs 0.001*cqr 0.1 0.01]; R2b=0;
while ( stol >= 0.00001 )
  if pqr==0
    [f,p,kvg,itr,corp,covp,covr,std,z,R2]=leasqr(x2,y,pin,"BC",stol,20,wt,dp);
  else
    [f,p,kvg,itr,corp,covp,covr,std,z,R2]=leasqr(x2,y,pin,"BC2",stol,20,wt,dp);
  end
  if kvg == 0, p=pin;, R2=R2b;, stol=0;, end
  stol=stol/10; pin=p; R2b=R2;
endwhile
if pqr==1, p(2)=abs(p(2));, end
qs=p(1), qr=p(2), hb=p(3), lambda=p(4), R2
if adv==1,
  CorrelationMatrix = corp
  StandardDeviation = sqrt(diag(covp))
end

# van Genuchten Model
disp ("=== VG model ===");
function ret=VG(x,p)
  n=abs(p(4))+1;
  ret = p(2)+(p(1)-p(2)).*(1+(abs(p(3)).*x).^n).^(1/n-1);
endfunction
function ret=VG2(x,p)
  n=abs(p(4))+1;
  ret = abs(p(2))+(p(1)-abs(p(2))).*(1+(abs(p(3)).*x).^n).^(1/n-1);
endfunction
qs=qsin; qr=qrin;
hm = x(sum(y > max(y)/2 + min(y)/2)+1);
alpha=1/hm; n=lambda+1;
if n < 1.1, n=1.1;, end
if n > 10, n=10;, end
pin=[qs, qr, alpha, n-1];
stol=1; dp=[0.00001*cqs 0.001*cqr 0.1 0.001]; R2b=0;
while ( stol >= 0.00001 )
  if pqr==0
    [f,p,kvg,itr,corp,covp,covr,std,Z,R2]=leasqr(x,y,pin,"VG",stol,20,wt,dp);
  else
    [f,p,kvg,itr,corp,covp,covr,std,Z,R2]=leasqr(x,y,pin,"VG2",stol,20,wt,dp);
  end
  if kvg == 0, p=pin;, R2=R2b;, stol=0;, end
  if imag(R2)^2>0, p=pin; R2=R2b;, stol=0;, end
  stol=stol/10; pin=p; R2b=R2;
endwhile
if pqr==1, p(2)=abs(p(2));, end
qs=p(1), qr=p(2), alpha=abs(p(3)), n=abs(p(4))+1, R2
if adv==1,
  CorrelationMatrix = corp
  StandardDeviation = sqrt(diag(covp))
end

# Lognormal Pore-Size Distribution Model of Kosugi
disp ("=== LN model ===");
function ret=LN(x,p)
  ret = p(2)+(p(1)-p(2)).*(1-normcdf((log(x/abs(p(3))))./abs(p(4))));
endfunction
function ret=LN2(x,p)
  ret = abs(p(2))+(p(1)-abs(p(2))).*(1-normcdf((log(x/abs(p(3))))./abs(p(4))));
endfunction
qs=qsin; qr=qrin; hm=1/alpha;
sigma=1.2*(n-1)^(-0.8);
if sigma < 0.15, sigma=0.15;, end
if sigma > 3, sigma=3;, end
pin=[qs, qr, hm, sigma];
stol=1; dp=[0.00001*cqs 0.0001*cqr 0.01 0.01]; R2b=0;
while ( stol >= 0.00001 )
  if pqr==0
    [f,p,kvg,itr,corp,covp,covr,std,Z,R2]=leasqr(x,y,pin,"LN",stol,20,wt,dp);
  else
    [f,p,kvg,itr,corp,covp,covr,std,Z,R2]=leasqr(x,y,pin,"LN2",stol,20,wt,dp);
  end
  if kvg == 0, p=pin;, R2=R2b;, stol=0;, end
  if imag(R2)^2>0, p=pin; R2=R2b;, stol=0;, end
  stol=stol/10; pin=p; R2b=R2;
endwhile
if pqr==1, p(2)=abs(p(2));, end
qs=p(1), qr=p(2), hm=abs(p(3)), sigma=abs(p(4)), R2
if adv==1,
  CorrelationMatrix = corp
  StandardDeviation = sqrt(diag(covp))
end


bimodal.m
#!/usr/bin/octave -qf
# SWRC Fit - bimodal.m
# URL: http://swrcfit.sourceforge.net/
# Author: Katsutoshi Seki
# License: GNU Public License
# Version: 1.2

# Read data from file
load swrc.txt
[s, i] = sort (swrc (:, 1));
sorted = swrc (i, :);
x=sorted(:,1);
y=sorted(:,2);
if columns(sorted)==2,
  wt=ones(length(y),1);
else
  wt=sorted(:,3);
end

# Setting
output_precision=7; # precision of the output
qsin = max(y); # initial value of qs
cqs=1; # cqs=1; qs is variable, cqs=0; qs is constant
qrin = min(y); # initial value of qr
cqr=1; # cqr=1; qr is variable, cqr=0; qr is constant
# qrin=0; cqr=0; # For setting qr=0 as a constant
pqr=1; # pqr=1; qr >= 0, pqr=0; qr can be negative
adv=0; # adv=1; advanced output; adv=0; normal output;

# Check if there are more than 7 points
if i < 8, disp("Too few points"), exit, end

# Bimodal model of Durner
disp ("=== DB model ===");

i = 3;
while (i < length(x)-1)
xi=x(1:i); yi=y(1:i);
qs=qsin; qr=min(yi);
S=(yi.-qr)./(qs-qr);
function ret=VG(x,p)
  ret = p(2)+(p(1)-p(2)).*((1+(abs(p(3)).*x).^(abs(p(4))+1)).^(1/(abs(p(4))+1)-1));
endfunction
hm = xi(sum (S > 0.5)+1);
alpha=1/hm; nn=1; pin=[1 0 alpha nn];
dp=[0 0 0.001 0.001]; stol=1; R2b=0;
wt2=ones(length(yi),1);
while ( stol >= 0.0001 )
  [f,p,kvg]=leasqr(xi,S,pin,"VG",0.01,20,wt2,dp);
  if kvg ==0, p=pin;, stol=0;, end
  stol=stol/10; pin=p;
endwhile
alpha1(i)=p(3); n1(i)=abs(p(4))+1;
y2=y.-(qs-qr).*(1+(alpha1(i).*x).^n1(i)).^(1./n1(i)-1);
hm = x(sum (S > 0.5)+1);
pin=[qr qrin 1/hm 1]; stol=1; 
dp=[0 0.0001*cqr 0.0001 0.0001]; R2b=0;
while ( stol >= 0.0001 )
  [f,p,kvg,itr,corp,covp,cov,std,Z,R2]=leasqr(x,y2,pin,"VG",stol,20,wt,dp);
  if kvg == 0, p=pin;, R2=R2b;, stol=0;, end
  if imag(R2)^2>0, p=pin; R2=R2b;, stol=0;, end
  stol=stol/10; pin=p; R2b=R2;
endwhile
alpha2(i)=real(p(3)); n2(i)=abs(real(p(4)))+1;
tr(i)=real(p(2)); w1(i)=(qs-qr)/(qs-tr(i));
R(i)=real(R2);
i++;
endwhile

# Set initial condition
[Rmax, i]=max(R);
pin=[qsin tr(i) w1(i) 1/alpha1(i) n1(i) 1/alpha2(i) n2(i)];

# Calculate
function ret=DB(x,p)
  n1=mod(abs(p(5)),49)+1; n2=mod(abs(p(7)),49)+1; w1=mod(p(3),1);
  ret = p(2)+(p(1)-p(2)).*(w1.*(1+(x./abs(p(4))).^n1).^(1/n1-1) + (1-w1).*(1+(x./abs(p(6))).^n2).^(1/n2-1));
endfunction
function ret=DB2(x,p)
  n1=mod(abs(p(5)),49)+1; n2=mod(abs(p(7)),49)+1; w1=mod(p(3),1);
  ret = abs(p(2))+(p(1)-abs(p(2))).*(w1.*(1+(x./abs(p(4))).^n1).^(1/n1-1) + (1-w1).*(1+(x./abs(p(6))).^n2).^(1/n2-1));
endfunction
stol=8; R2b=0;
dp=[0 0.001*cqr 0.001 0 0.01 0 0.01];
while ( stol >= 0.5 )
  [f,p,kvg,itr,corp,covp,covr,std,Z,R2]=leasqr(x,y,pin,"DB",stol,20,wt,dp);
  if kvg == 0, p=pin;, R2=R2b;, stol=0;, end
  if imag(R2)^2>0, p=pin; R2=R2b;, stol=0;, end
  stol=stol/4; pin=p; R2b=R2;
endwhile
stol=1; R2b=0;
dp=[0.000001*cqs 0.001*cqr 0.001 0.01 0.001 0.1 0.001];
while ( stol >= 0.00001 )
  if pqr==0
    [f,p,kvg,itr,corp,covp,covr,std,Z,R2]=leasqr(x,y,pin,"DB",stol,20,wt,dp);
  else
    [f,p,kvg,itr,corp,covp,covr,std,Z,R2]=leasqr(x,y,pin,"DB2",stol,20,wt,dp);
  end
  if kvg == 0, p=pin;, R2=R2b;, stol=0;, end
  if imag(R2)^2>0, p=pin; R2=R2b;, stol=0;, end
  stol=stol/8; pin=p; R2b=R2;
endwhile
if abs(p(4)) > abs(p(6))
  p=[p(1) p(2) 1-p(3) p(6) p(7) p(4) p(5)];
end
w1 = mod(p(3),1);
if pqr==1, p(2)=abs(p(2));, end
if (w1 > 0.01) * (w1 < 0.99)
  qs = p(1)
  qr = p(2)
  w1
  alpha(1) = 1/abs(p(4));
  alpha1=alpha(1)
  n(1) = mod(abs(p(5)),49)+1;
  n1=n(1)
 alpha(2) = 1/abs(p(6));
 alpha2=alpha(2)
 n(2) = mod(abs(p(7)),49)+1;
 n2=n(2)
  R2
  if adv==1,
    CorrelationMatrix = corp
    StandardDeviation = sqrt(diag(covp))
  end
else
  disp ("Not bimodal");
end

# Bimodal Lognormal model
disp ("=== BL model ===");

# Initial condition
qs=p(1); qr=p(2);
i=1;
while (i<3)
v=[-1:0.2:4]'; xx=10.^v; m=1-1/n(i); yy=(1.+(alpha(i).*xx).^n(i)).^(-m);
sigma=(n(i)-1)^(-2/3);
pin=[1 0 1/alpha(i) sigma];
wt3=ones(length(xx),1);
dp=[0 0 0.1 0.01];
function ret=LN(x,p)
  ret = p(2)+(p(1)-p(2)).*(1-normcdf((log(x/abs(p(3))))./abs(p(4))));
endfunction
function ret=LN2(x,p)
  ret = abs(p(2))+(p(1)-abs(p(2))).*(1-normcdf((log(x/abs(p(3))))./abs(p(4))));
endfunction
[f,p]=leasqr(xx,yy,pin,"LN",1,20,wt3,dp);
hm(i)=abs(p(3)); sigma(i)=abs(p(4));
i++;
endwhile
pin=[qs, qr, w1, hm(1), sigma(1), hm(2), sigma(2)];

# Calculate
function ret=BL(x,p)
  sigma1=mod(abs(p(5)),50)+0.05; sigma2=mod(abs(p(7)),50)+0.05; w1=mod(p(3),1);
  ret = p(2)+(p(1)-p(2)).*(w1.*(1-normcdf((log(x./abs(p(4)))./sigma1))) + (1-w1).*(1-normcdf((log(x./abs(p(6)))./sigma2))));
endfunction
function ret=BL2(x,p)
  sigma1=mod(abs(p(5)),50)+0.05; sigma2=mod(abs(p(7)),50)+0.05; w1=mod(p(3),1);
  ret = abs(p(2))+(p(1)-abs(p(2))).*(w1.*(1-normcdf((log(x./abs(p(4)))./sigma1))) + (1-w1).*(1-normcdf((log(x./abs(p(6)))./sigma2))));
endfunction
stol=8; R2b=0;
dp=[0.000001*cqs 0.05*cqr 0.0001 0.001 0.001 0.001 0.001];
while ( stol >= 0.00001 )
  if pqr==0
    [f,p,kvg,itr,corp,covp,covr,std,Z,R2]=leasqr(x,y,pin,"BL",stol,20,wt,dp);
  else
    [f,p,kvg,itr,corp,covp,covr,std,Z,R2]=leasqr(x,y,pin,"BL2",stol,20,wt,dp);
  end
  if kvg == 0, p=pin;, R2=R2b;, stol=0;, end
  if imag(R2)^2>0, p=pin; R2=R2b;, stol=0;, end
  stol=stol/8; pin=p; R2b=R2;
endwhile
if abs(p(4)) > abs(p(6))
  p=[p(1) p(2) 1-p(3) p(6) p(7) p(4) p(5)];
end

w1 = mod(p(3),1);
if pqr==1, p(2)=abs(p(2));, end
if (w1 > 0.01) * (w1 < 0.99) +1
  qs = p(1)
  qr = p(2)
  w1
  hm1 = abs(p(4))
  sigma1 = mod(abs(p(5)),50)+0.05
  hm2 = abs(p(6))
  sigma2 = mod(abs(p(7)),50)+0.05
  R2
  if adv==1,
    CorrelationMatrix = corp
    StandardDeviation = sqrt(diag(covp))
  end
else
  disp ("Not bimodal");
end



swrcfit.m
#!/usr/bin/octave -qf
# swrcfit - Soil Water Retention Curve Fit
# URL: http://swrcfit.sourceforge.net/
# Author: Katsutoshi Seki (http://researchmap.jp/sekik/)
# Licence: GNU Public License
# Version: 1.3
#
# Usage: swrcfit DataFilename [SettingFilename]

# Read commandline parameter

argc = length (argv);
if argc == 0,
  disp ("Usage: swrcfit DataFilename [SettingFilename]");
  disp ("See https://github.com/sekika/swrcfit/blob/master/README.md");
  return;
end

arg_list = argv ();
filename = arg_list{1};

setting = "";
if argc > 1, setting = arg_list{2}; end

# Read data from file
swrc = load ("-ascii",filename);
[s, i] = sort (swrc (:, 1));
sorted = swrc (i, :);
x=sorted(:,1); y=sorted(:,2);
if columns(sorted)==2,
  wt=ones(length(y),1);
else
  wt=sorted(:,3);
end

# Default Setting
mode = 1; # Unimodal model
output_precision(5); # precision of the output
qsin = max(y); # initial value of qs
cqs=1; # cqs=1; qs is variable, cqs=0; qs is constant
qrin = min(y); # initial value of qr
cqr=1; # cqr=1; qr is variable, cqr=0; qr is constant
# qrin=0; cqr=0; # For setting qr=0 as a constant
pqr=1; # pqr=1; qr >= 0, pqr=0; qr can be negative
adv=0; # adv=1; advanced output; adv=0; normal output;

if (argc > 1), source (setting) end

######################### Unimodal model #####################
if (mode - 2 ~= 0),

# Set initial values
qs=qsin; qr=qrin;
hbi = sum (y > max(y) * 0.95 + min(y) * 0.05)+1;
hb=x(hbi); hby=(y(hbi)-qr)/(qs-qr);
if hb==0, hb=x(2)/10;, end
hli = sum (y > max(y) * 0.15 + min(y) * 0.85);
if hbi==hli; hli=hli+1;, end
hl=x(hli); hly=(y(hli)-qr)/(qs-qr);
lambda=-log(hly/hby)/log(hl/hb);
if lambda < 0.1, lambda=0.1;, end
if lambda > 10, lambda=10;, end

# Brooks and Corey Model
disp ("=== BC model ===");
x2 = x .+ (x == 0) .* 0.001;
function ret=BC(x,p)
  ret = (x > p(3)) .* (p(2)+(p(1)-p(2)).*(x./p(3)).^(-p(4))) + (x <= p(3)) .* p(1);
endfunction
function ret=BC2(x,p)
  ret = (x > p(3)) .* (abs(p(2))+(p(1)-abs(p(2))).*(x./p(3)).^(-p(4))) + (x <= p(3)) .* p(1);
endfunction

pin=[qs, qr, hb, lambda];
stol=1;
dp=[0.00001*cqs 0.001*cqr 0.1 0.01]; R2b=0;
while ( stol >= 0.00001 )
  if pqr==0
    [f,p,kvg,itr,corp,covp,covr,std,z,R2]=leasqr(x2,y,pin,"BC",stol,20,wt,dp);
  else
    [f,p,kvg,itr,corp,covp,covr,std,z,R2]=leasqr(x2,y,pin,"BC2",stol,20,wt,dp);
  end
  if kvg == 0, p=pin;, R2=R2b;, stol=0;, end
  stol=stol/10; pin=p; R2b=R2;
endwhile
if pqr==1, p(2)=abs(p(2));, end
qs=p(1), qr=p(2), hb=p(3), lambda=p(4), R2
if adv==1,
  CorrelationMatrix = corp
  StandardDeviation = sqrt(diag(covp))
end

# van Genuchten Model
disp ("=== VG model ===");
function ret=VG(x,p)
  n=abs(p(4))+1;
  ret = p(2)+(p(1)-p(2)).*(1+(abs(p(3)).*x).^n).^(1/n-1);
endfunction
function ret=VG2(x,p)
  n=abs(p(4))+1;
  ret = abs(p(2))+(p(1)-abs(p(2))).*(1+(abs(p(3)).*x).^n).^(1/n-1);
endfunction
qs=qsin; qr=qrin;
hm = x(sum(y > max(y)/2 + min(y)/2)+1);
alpha=1/hm; n=lambda+1;
if n < 1.1, n=1.1;, end
if n > 10, n=10;, end
pin=[qs, qr, alpha, n-1];
stol=1; dp=[0.00001*cqs 0.001*cqr 0.1 0.001]; R2b=0;
while ( stol >= 0.00001 )
  if pqr==0
    [f,p,kvg,itr,corp,covp,covr,std,Z,R2]=leasqr(x,y,pin,"VG",stol,20,wt,dp);
  else
    [f,p,kvg,itr,corp,covp,covr,std,Z,R2]=leasqr(x,y,pin,"VG2",stol,20,wt,dp);
  end
  if kvg == 0, p=pin;, R2=R2b;, stol=0;, end
  if imag(R2)^2>0, p=pin; R2=R2b;, stol=0;, end
  stol=stol/10; pin=p; R2b=R2;
endwhile
if pqr==1, p(2)=abs(p(2));, end
qs=p(1), qr=p(2), alpha=abs(p(3)), n=abs(p(4))+1, R2
if adv==1,
  CorrelationMatrix = corp
  StandardDeviation = sqrt(diag(covp))
end

# Lognormal Pore-Size Distribution Model of Kosugi
disp ("=== LN model ===");
function ret=LN(x,p)
  ret = p(2)+(p(1)-p(2)).*(1-normcdf((log(x/abs(p(3))))./abs(p(4))));
endfunction
function ret=LN2(x,p)
  ret = abs(p(2))+(p(1)-abs(p(2))).*(1-normcdf((log(x/abs(p(3))))./abs(p(4))));
endfunction
qs=qsin; qr=qrin; hm=1/alpha;
sigma=1.2*(n-1)^(-0.8);
if sigma < 0.15, sigma=0.15;, end
if sigma > 3, sigma=3;, end
pin=[qs, qr, hm, sigma];
stol=1; dp=[0.00001*cqs 0.0001*cqr 0.01 0.01]; R2b=0;
while ( stol >= 0.00001 )
  if pqr==0
    [f,p,kvg,itr,corp,covp,covr,std,Z,R2]=leasqr(x,y,pin,"LN",stol,20,wt,dp);
  else
    [f,p,kvg,itr,corp,covp,covr,std,Z,R2]=leasqr(x,y,pin,"LN2",stol,20,wt,dp);
  end
  if kvg == 0, p=pin;, R2=R2b;, stol=0;, end
  if imag(R2)^2>0, p=pin; R2=R2b;, stol=0;, end
  stol=stol/10; pin=p; R2b=R2;
endwhile
if pqr==1, p(2)=abs(p(2));, end
qs=p(1), qr=p(2), hm=abs(p(3)), sigma=abs(p(4)), R2
if adv==1,
  CorrelationMatrix = corp
  StandardDeviation = sqrt(diag(covp))
end

end

################### Bimodal model #####################
if (mode > 1),

# Check if there are more than 7 points
if i < 8, disp("Too few points for bimodal analysis."), exit, end

# Bimodal model of Durner
disp ("=== DB model ===");

i = 3;
while (i < length(x)-1)
xi=x(1:i); yi=y(1:i);
qs=qsin; qr=min(yi);
S=(yi.-qr)./(qs-qr);
function ret=VG(x,p)
  ret = p(2)+(p(1)-p(2)).*((1+(abs(p(3)).*x).^(abs(p(4))+1)).^(1/(abs(p(4))+1)-1));
endfunction
hm = xi(sum (S > 0.5)+1);
alpha=1/hm; nn=1; pin=[1 0 alpha nn];
dp=[0 0 0.001 0.001]; stol=1; R2b=0;
wt2=ones(length(yi),1);
while ( stol >= 0.0001 )
  [f,p,kvg]=leasqr(xi,S,pin,"VG",0.01,20,wt2,dp);
  if kvg ==0, p=pin;, stol=0;, end
  stol=stol/10; pin=p;
endwhile
alpha1(i)=p(3); n1(i)=abs(p(4))+1;
y2=y.-(qs-qr).*(1+(alpha1(i).*x).^n1(i)).^(1./n1(i)-1);
hm = x(sum (S > 0.5)+1);
pin=[qr qrin 1/hm 1]; stol=1; 
dp=[0 0.0001*cqr 0.0001 0.0001]; R2b=0;
while ( stol >= 0.0001 )
  [f,p,kvg,itr,corp,covp,cov,std,Z,R2]=leasqr(x,y2,pin,"VG",stol,20,wt,dp);
  if kvg == 0, p=pin;, R2=R2b;, stol=0;, end
  if imag(R2)^2>0, p=pin; R2=R2b;, stol=0;, end
  stol=stol/10; pin=p; R2b=R2;
endwhile
alpha2(i)=real(p(3)); n2(i)=abs(real(p(4)))+1;
tr(i)=real(p(2)); w1(i)=(qs-qr)/(qs-tr(i));
R(i)=real(R2);
i++;
endwhile

# Set initial condition
[Rmax, i]=max(R);
pin=[qsin tr(i) w1(i) 1/alpha1(i) n1(i) 1/alpha2(i) n2(i)];

# Calculate
function ret=DB(x,p)
  n1=mod(abs(p(5)),49)+1; n2=mod(abs(p(7)),49)+1; w1=mod(p(3),1);
  ret = p(2)+(p(1)-p(2)).*(w1.*(1+(x./abs(p(4))).^n1).^(1/n1-1) + (1-w1).*(1+(x./abs(p(6))).^n2).^(1/n2-1));
endfunction
function ret=DB2(x,p)
  n1=mod(abs(p(5)),49)+1; n2=mod(abs(p(7)),49)+1; w1=mod(p(3),1);
  ret = abs(p(2))+(p(1)-abs(p(2))).*(w1.*(1+(x./abs(p(4))).^n1).^(1/n1-1) + (1-w1).*(1+(x./abs(p(6))).^n2).^(1/n2-1));
endfunction
stol=8; R2b=0;
dp=[0 0.001*cqr 0.001 0 0.01 0 0.01];
while ( stol >= 0.5 )
  [f,p,kvg,itr,corp,covp,covr,std,Z,R2]=leasqr(x,y,pin,"DB",stol,20,wt,dp);
  if kvg == 0, p=pin;, R2=R2b;, stol=0;, end
  if imag(R2)^2>0, p=pin; R2=R2b;, stol=0;, end
  stol=stol/4; pin=p; R2b=R2;
endwhile
stol=1; R2b=0;
dp=[0.000001*cqs 0.001*cqr 0.001 0.01 0.001 0.1 0.001];
while ( stol >= 0.00001 )
  if pqr==0
    [f,p,kvg,itr,corp,covp,covr,std,Z,R2]=leasqr(x,y,pin,"DB",stol,20,wt,dp);
  else
    [f,p,kvg,itr,corp,covp,covr,std,Z,R2]=leasqr(x,y,pin,"DB2",stol,20,wt,dp);
  end
  if kvg == 0, p=pin;, R2=R2b;, stol=0;, end
  if imag(R2)^2>0, p=pin; R2=R2b;, stol=0;, end
  stol=stol/8; pin=p; R2b=R2;
endwhile
if abs(p(4)) > abs(p(6))
  p=[p(1) p(2) 1-p(3) p(6) p(7) p(4) p(5)];
end
w1 = mod(p(3),1);
if pqr==1, p(2)=abs(p(2));, end
if (w1 > 0.01) * (w1 < 0.99)
  qs = p(1)
  qr = p(2)
  w1
  alpha(1) = 1/abs(p(4));
  alpha1=alpha(1)
  n(1) = mod(abs(p(5)),49)+1;
  n1=n(1)
 alpha(2) = 1/abs(p(6));
 alpha2=alpha(2)
 n(2) = mod(abs(p(7)),49)+1;
 n2=n(2)
  R2
  if adv==1,
    CorrelationMatrix = corp
    StandardDeviation = sqrt(diag(covp))
  end
else
  disp ("Not bimodal");
end

# Bimodal Lognormal model
disp ("=== BL model ===");

# Initial condition
qs=p(1); qr=p(2);
i=1;
while (i<3)
v=[-1:0.2:4]'; xx=10.^v; m=1-1/n(i); yy=(1.+(alpha(i).*xx).^n(i)).^(-m);
sigma=(n(i)-1)^(-2/3);
pin=[1 0 1/alpha(i) sigma];
wt3=ones(length(xx),1);
dp=[0 0 0.1 0.01];
function ret=LN(x,p)
  ret = p(2)+(p(1)-p(2)).*(1-normcdf((log(x/abs(p(3))))./abs(p(4))));
endfunction
function ret=LN2(x,p)
  ret = abs(p(2))+(p(1)-abs(p(2))).*(1-normcdf((log(x/abs(p(3))))./abs(p(4))));
endfunction
[f,p]=leasqr(xx,yy,pin,"LN",1,20,wt3,dp);
hm(i)=abs(p(3)); sigma(i)=abs(p(4));
i++;
endwhile
pin=[qs, qr, w1, hm(1), sigma(1), hm(2), sigma(2)];

# Calculate
function ret=BL(x,p)
  sigma1=mod(abs(p(5)),50)+0.05; sigma2=mod(abs(p(7)),50)+0.05; w1=mod(p(3),1);
  ret = p(2)+(p(1)-p(2)).*(w1.*(1-normcdf((log(x./abs(p(4)))./sigma1))) + (1-w1).*(1-normcdf((log(x./abs(p(6)))./sigma2))));
endfunction
function ret=BL2(x,p)
  sigma1=mod(abs(p(5)),50)+0.05; sigma2=mod(abs(p(7)),50)+0.05; w1=mod(p(3),1);
  ret = abs(p(2))+(p(1)-abs(p(2))).*(w1.*(1-normcdf((log(x./abs(p(4)))./sigma1))) + (1-w1).*(1-normcdf((log(x./abs(p(6)))./sigma2))));
endfunction
stol=8; R2b=0;
dp=[0.000001*cqs 0.05*cqr 0.0001 0.001 0.001 0.001 0.001];
while ( stol >= 0.00001 )
  if pqr==0
    [f,p,kvg,itr,corp,covp,covr,std,Z,R2]=leasqr(x,y,pin,"BL",stol,20,wt,dp);
  else
    [f,p,kvg,itr,corp,covp,covr,std,Z,R2]=leasqr(x,y,pin,"BL2",stol,20,wt,dp);
  end
  if kvg == 0, p=pin;, R2=R2b;, stol=0;, end
  if imag(R2)^2>0, p=pin; R2=R2b;, stol=0;, end
  stol=stol/8; pin=p; R2b=R2;
endwhile
if abs(p(4)) > abs(p(6))
  p=[p(1) p(2) 1-p(3) p(6) p(7) p(4) p(5)];
end

w1 = mod(p(3),1);
if pqr==1, p(2)=abs(p(2));, end
if (w1 > 0.01) * (w1 < 0.99) +1
  qs = p(1)
  qr = p(2)
  w1
  hm1 = abs(p(4))
  sigma1 = mod(abs(p(5)),50)+0.05
  hm2 = abs(p(6))
  sigma2 = mod(abs(p(7)),50)+0.05
  R2
  if adv==1,
    CorrelationMatrix = corp
    StandardDeviation = sqrt(diag(covp))
  end
else
  disp ("Not bimodal");
end

end
